src/main/java/com/google/cloud/tools/app/ProcessCaller.java

```diff
@@ -0,0 +1,189 @@
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, true);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      File workingDirectory) {
+    this(tool, arguments, flags, workingDirectory, true);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      boolean synchronous) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, synchronous);
+  }
+
+  public ProcessCaller(Tool tool, Collection<String> arguments, Map<Option, String> flags,
+      File workingDirectory, boolean synchronous) {
+    this.workingDirectory = workingDirectory;
+    this.synchronous = synchronous;
+    this.command = prepareCommand(tool, arguments, flags);
+  }
+
+  public boolean call() throws GCloudExecutionException {
+    LOG.info("Calling " + Joiner.on(" ").join(command));
+
+    ProcessBuilder processBuilder = new ProcessBuilder(command);
+    processBuilder.directory(workingDirectory);
+    processBuilder.inheritIO();
+    try {
+      Process gcloudProcess = processBuilder.start();
+
+      // If call is synchronous, wait for the process to end.
+      if (synchronous) {
+        int exitStatus = gcloudProcess.waitFor();
+
+        if (exitStatus != 0) {
+          throw new GCloudExecutionException(exitStatus);
+        }
+      }
+    } catch (IOException|InterruptedException ex) {
+      LOG.severe("Error running gcloud CLI. " + ex);
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Finds the executable path for gcloud.
+   */
+  @VisibleForTesting
+  public String getGcloudPath() {
+    String gcloudLocation = cloudSdkLocation + "/bin/gcloud";
+    if (!new File(gcloudLocation).exists()) {
+      throw new RuntimeException("Could not locate gcloud from Cloud SDK directory \""
+          + cloudSdkLocation + "/bin\". Please provide the correct Cloud SDK root directory.");
+    }
+
+    return gcloudLocation;
+  }
+
+  /**
+   * Finds the executable path for dev_appserver.py.
+   */
+  protected String getDevAppserverPath() {
+    String devAppserverLocation = cloudSdkLocation + "/bin/dev_appserver.py";
+    if (!new File(devAppserverLocation).exists()) {
+      throw new RuntimeException("Could not locate dev_appserver from Cloud SDK directory \""
+          + cloudSdkLocation + "/bin\". Please provide the correct Cloud SDK root directory.");
+    }
+
+    return devAppserverLocation;
+  }
+
+  @VisibleForTesting
+  public List<String> getCommand() {
+    return command;
+  }
+
+  /**
+   * Prepares the gcloud command ran by the {@link Action}.
+   *
+   * <p>Obsoleted when gcloud CLI is no longer a dependency.
+   */
+  protected List<String> prepareCommand(Tool tool, Collection<String> arguments,
+      Map<Option, String> flags) {
+    List<String> command = new ArrayList<>();
+
+    if (tool.equals(Tool.DEV_APPSERVER)) {
```

Comment:
It's an anti pattern in languages like java to use a Flag value like Tool to branch within a method.  One way to address it might be to have the Tool object provide it's command prefix.